src/VecSim/index_factories/hnsw_disk_factory.cpp

```diff
@@ -0,0 +1,303 @@
+inline VecSimIndex *NewDiskIndex_ChooseMultiOrSingle(std::ifstream &input,
+                                                     const HNSWParams *params,
+                                                     const AbstractIndexInitParams &abstractInitParams,
+                                                     IndexComponents<DataType, DistType> &components,
+                                                     rocksdb::DB *db,
+                                                     rocksdb::ColumnFamilyHandle *cf,
+                                                     Serializer::EncodingVersion version) {
+    // Note: HNSWDiskIndex doesn't have separate Single/Multi variants like in-memory HNSW
+    // It handles both cases internally based on params->multi
+    return new (abstractInitParams.allocator) HNSWDiskIndex<DataType, DistType>(
+        input, params, abstractInitParams, components, db, cf, version);
+}
+
+// Initialize @params from file (same format as regular HNSW)
+static void InitializeParams(std::ifstream &source_params, HNSWParams &params) {
+    Serializer::readBinaryPOD(source_params, params.dim);
+    Serializer::readBinaryPOD(source_params, params.type);
+    Serializer::readBinaryPOD(source_params, params.metric);
+    Serializer::readBinaryPOD(source_params, params.blockSize);
+    Serializer::readBinaryPOD(source_params, params.multi);
+    Serializer::readBinaryPOD(source_params, params.initialCapacity);
+}
+
+std::string GetCheckpointDir(const std::string &folder_path) {
+    std::filesystem::path path(folder_path);
+    return (path / "rocksdb").string();
+}
+
+VecSimIndex *NewIndex(const VecSimParams *params) {
+    const HNSWDiskParams *hnswDiskParams = &params->algoParams.hnswDiskParams;
+    const HNSWParams *hnswParams = &params->algoParams.hnswParams;
+
+    // Get the dbPath from params
+    std::string dbPath = std::string(hnswDiskParams->dbPath);
+    std::cerr << "RocksDB path: " << dbPath << std::endl;
+
+    // Ensure the directory exists
+    std::filesystem::create_directories(dbPath);
+
+    // Open RocksDB
+    rocksdb::Options options;
+    options.create_if_missing = true;
+    options.error_if_exists = false;
+    options.statistics = rocksdb::CreateDBStatistics();
+
+    rocksdb::DB *db_ptr = nullptr;
+    rocksdb::Status status = rocksdb::DB::Open(options, dbPath, &db_ptr);
+    if (!status.ok()) {
+        throw std::runtime_error("Failed to open RocksDB: " + status.ToString());
+    }
+
+    // Store in RAII wrapper (will close DB on exit, but won't delete directory)
+    managed_rocksdb = std::make_unique<ManagedRocksDB>(db_ptr, dbPath);
+
+    // Create AbstractIndexInitParams
+    AbstractIndexInitParams abstractInitParams = NewAbstractInitParams(params);
+
+    // Create the appropriate disk index based on data type
+    if (hnswParams->type == VecSimType_FLOAT32) {
+        IndexComponents<float, float> indexComponents = CreateIndexComponents<float, float>(
```

Comment:
```suggestion
        IndexComponents<float, float> indexComponents = CreateIndexComponents<float, float>(
```
        IndexComponents<float, float> indexComponents = CreateQuantiaedIndexComponents<float, float>(